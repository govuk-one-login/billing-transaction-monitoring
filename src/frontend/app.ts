import express from "express";
import nunjucks from "nunjucks";
import path from "path";
import { fileURLToPath } from "url";

let dirname;
try {
  dirname = __dirname;
} catch {
  const filePath = fileURLToPath(import.meta.url);
  dirname = path.dirname(filePath);
}

const app = express();

const viewDir = path.join(dirname, "views");

nunjucks.configure(["node_modules/govuk-frontend/", viewDir], {
  autoescape: true,
  express: app,
});

app.engine("njk", nunjucks.render);

app.get("/", (_, response) => {
  response.render("index.njk");
});

app.use("/images", express.static(dirname));
app.use("/scripts", express.static(dirname));
app.use("/styles", express.static(path.join(dirname, "./assets/styles")));

console.log("boab?");

export { app };
