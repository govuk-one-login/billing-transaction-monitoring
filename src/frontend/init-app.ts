import express, { Express } from "express";
import nunjucks from "nunjucks";
import path from "path";
import { fileURLToPath } from "url";
import "dotenv/config";
import { getHandler, getPagePath, PAGES } from "./pages";

let dirname: string;
try {
  dirname = __dirname;
} catch {
  const filePath = fileURLToPath(import.meta.url);
  dirname = path.dirname(filePath);
}

const shouldLoadFromNodeModules =
  process.env.NODE_ENV === "development" || process.env.NODE_ENV === "test";

export const initApp = (app: Express): void => {
  const viewDir = path.join(dirname, "views");

  const templatePath = shouldLoadFromNodeModules
    ? "node_modules/govuk-frontend"
    : dirname;

  nunjucks.configure([templatePath, viewDir], {
    autoescape: true,
    express: app,
  });

  app.engine("njk", nunjucks.render);

  PAGES.forEach((page) => {
    app.get(getPagePath(page), getHandler(page));
  });

  const assetsPath = shouldLoadFromNodeModules
    ? path.join("node_modules/govuk-frontend/govuk", "./assets")
    : path.join(dirname, "./assets");

  const scriptsPath = shouldLoadFromNodeModules
    ? "node_modules/govuk-frontend/govuk"
    : assetsPath;

  const stylesPath = shouldLoadFromNodeModules
    ? path.join(dirname, "./styles")
    : assetsPath;

  app.use("/assets", express.static(assetsPath));
  app.use("/assets/scripts", express.static(scriptsPath));
  app.use("/assets/styles", express.static(stylesPath));
};
