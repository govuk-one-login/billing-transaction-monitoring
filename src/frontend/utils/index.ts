export * from "./months";
export * from "./status-label";
export * from "./invoice-statuses";
export * from "./links";
export * from "./title";
