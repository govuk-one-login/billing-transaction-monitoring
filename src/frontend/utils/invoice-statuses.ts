import { StatusLabel } from "./status-label";

export enum InvoiceBannerClass {
  warning = "warning",
  error = "error",
  notice = "notice",
  payable = "payable",
}

export type InvoiceStatus = {
  bannerText: string;
  bannerClass: string;
  statusLabel: StatusLabel;
};

export enum InvoiceStatuses {
  invoiceAndEventsMissing,
  invoiceDataMissing,
  eventsMissing,
  unableToFindRate,
  invoiceHasUnexpectedCharge,
  invoiceAboveThreshold,
  invoiceBelowThreshold,
  invoiceWithinThreshold,
  invoiceHasNoCharge,
}

export const invoiceStatusLabelLookup: Record<InvoiceStatuses, StatusLabel> = {
  [InvoiceStatuses.invoiceAndEventsMissing]: {
    message: "Pending",
    class: "govuk-tag--blue",
  },
  [InvoiceStatuses.invoiceDataMissing]: {
    message: "Warning",
    class: "govuk-tag--blue",
  },

  [InvoiceStatuses.eventsMissing]: {
    message: "Error",
    class: "govuk-tag--red",
  },
  [InvoiceStatuses.unableToFindRate]: {
    message: "Error",
    class: "govuk-tag--red",
  },

  [InvoiceStatuses.invoiceHasUnexpectedCharge]: {
    message: "Warning",
    class: "govuk-tag--yellow",
  },
  [InvoiceStatuses.invoiceAboveThreshold]: {
    message: "Warning",
    class: "govuk-tag--yellow",
  },

  [InvoiceStatuses.invoiceBelowThreshold]: {
    message: "Below Threshold",
    class: "govuk-tag--green",
  },
  [InvoiceStatuses.invoiceWithinThreshold]: {
    message: "Within Threshold",
    class: "govuk-tag--green",
  },
  [InvoiceStatuses.invoiceHasNoCharge]: {
    message: "No Charge",
    class: "govuk-tag--green",
  },
};

export const invoiceStatusLookup: Record<InvoiceStatuses, InvoiceStatus> = {
  [InvoiceStatuses.invoiceAndEventsMissing]: {
    bannerText: "Invoice and events missing",
    bannerClass: InvoiceBannerClass.notice,
    statusLabel:
      invoiceStatusLabelLookup[InvoiceStatuses.invoiceAndEventsMissing],
  },
  [InvoiceStatuses.invoiceHasNoCharge]: {
    bannerText: "No charge",
    bannerClass: InvoiceBannerClass.payable,
    statusLabel: invoiceStatusLabelLookup[InvoiceStatuses.invoiceHasNoCharge],
  },
  [InvoiceStatuses.unableToFindRate]: {
    bannerText: "Unable to find rate",
    bannerClass: InvoiceBannerClass.error,
    statusLabel: invoiceStatusLabelLookup[InvoiceStatuses.unableToFindRate],
  },
  [InvoiceStatuses.invoiceDataMissing]: {
    bannerText: "Invoice data missing",
    bannerClass: InvoiceBannerClass.notice,
    statusLabel: invoiceStatusLabelLookup[InvoiceStatuses.invoiceDataMissing],
  },
  [InvoiceStatuses.eventsMissing]: {
    bannerText: "Events missing",
    bannerClass: InvoiceBannerClass.error,
    statusLabel: invoiceStatusLabelLookup[InvoiceStatuses.eventsMissing],
  },
  [InvoiceStatuses.invoiceHasUnexpectedCharge]: {
    bannerText: "Unexpected invoice charge",
    bannerClass: InvoiceBannerClass.warning,
    statusLabel:
      invoiceStatusLabelLookup[InvoiceStatuses.invoiceHasUnexpectedCharge],
  },
  [InvoiceStatuses.invoiceAboveThreshold]: {
    bannerText: "Invoice above threshold",
    bannerClass: InvoiceBannerClass.warning,
    statusLabel:
      invoiceStatusLabelLookup[InvoiceStatuses.invoiceAboveThreshold],
  },
  [InvoiceStatuses.invoiceBelowThreshold]: {
    bannerText: "Invoice below threshold",
    bannerClass: InvoiceBannerClass.payable,
    statusLabel:
      invoiceStatusLabelLookup[InvoiceStatuses.invoiceBelowThreshold],
  },
  [InvoiceStatuses.invoiceWithinThreshold]: {
    bannerText: "Invoice within threshold",
    bannerClass: InvoiceBannerClass.payable,
    statusLabel:
      invoiceStatusLabelLookup[InvoiceStatuses.invoiceWithinThreshold],
  },
};
