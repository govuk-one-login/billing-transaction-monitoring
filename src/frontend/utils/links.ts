import { getUrl, Page } from "../pages";

export type LinkData = {
  text: string;
  href: string;
};

export const getLinkData = async <TParams extends Record<string, string>>(
  page: Page<TParams, unknown>,
  requestParams: TParams
): Promise<LinkData> => ({
  text: await page.titleGetter(requestParams),
  href: getUrl(page, requestParams),
});
