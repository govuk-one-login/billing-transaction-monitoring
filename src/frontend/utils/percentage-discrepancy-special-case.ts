import { statusLabels } from "./status-label";

export enum InvoiceBannerStatus {
  invoiceAndEventsMissing = "Invoice and events missing",
  noCharge = "No charge",
  unableToFindRate = "Unable to find rate",
  invoiceDataMissing = "Invoice data missing",
  eventsMissing = "Events missing",
  unexpectedInvoiceCharge = "Unexpected invoice charge",
  invoiceAboveThreshold = "Invoice above threshold",
  invoiceBelowThreshold = "Invoice below threshold",
  invoiceWithinThreshold = "Invoice within threshold",
}

export enum InvoiceBannerClass {
  warning = "warning",
  error = "error",
  notice = "notice",
  payable = "payable",
}

export const percentageDiscrepancySpecialCase = {
  MN_NO_CHARGE: {
    magicNumber: "-1234567.01",
    bannerText: InvoiceBannerStatus.noCharge,
    statusLabel: statusLabels.STATUS_LABEL_WITHIN_THRESHOLD,
  },
  MN_RATES_MISSING: {
    magicNumber: "-1234567.02",
    bannerText: InvoiceBannerStatus.unableToFindRate,
    statusLabel: statusLabels.STATUS_LABEL_ERROR,
  },
  MN_INVOICE_MISSING: {
    magicNumber: "-1234567.03",
    bannerText: InvoiceBannerStatus.invoiceDataMissing,
    statusLabel: statusLabels.STATUS_LABEL_PENDING,
  },
  MN_EVENTS_MISSING: {
    magicNumber: "-1234567.04",
    bannerText: InvoiceBannerStatus.eventsMissing,
    statusLabel: statusLabels.STATUS_LABEL_ERROR,
  },
  MN_UNEXPECTED_CHARGE: {
    magicNumber: "-1234567.05",
    bannerText: InvoiceBannerStatus.unexpectedInvoiceCharge,
    statusLabel: statusLabels.STATUS_LABEL_ERROR,
  },
};
