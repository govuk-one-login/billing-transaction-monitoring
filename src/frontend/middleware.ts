import crypto from "crypto";
import helmet from "helmet";
import express, { ErrorRequestHandler, RequestHandler } from "express";
import { shouldLoadFromNodeModules } from "./utils/should-load-from-node-modules";
import path from "path";
import { rootDir } from "./utils/root-dir";
import { logger } from "../shared/utils";

export type Middlewares = Array<{ handler: RequestHandler; route?: string }>;

const generateNonce: RequestHandler = (_, response, next) => {
  response.locals.cspNonce = crypto.randomBytes(16).toString("hex");
  next();
};

const generateCSP: RequestHandler = (_, response, next) => {
  helmet({
    contentSecurityPolicy: {
      directives: {
        "connect-src": "https://accounts.google.com/gsi/",
        "frame-src": "https://accounts.google.com/gsi/",
        "script-src": ["'self'", `'nonce-${response.locals.cspNonce}'`],
        "style-src": ["'self'", "https://accounts.google.com/gsi/style"],
      },
    },
  });
  next();
};

const assetsPath = shouldLoadFromNodeModules
  ? path.join("node_modules/govuk-frontend/govuk", "./assets")
  : path.join(rootDir, "./assets");
const staticAssets = express.static(assetsPath);

const scriptsPath = shouldLoadFromNodeModules
  ? "node_modules/govuk-frontend/govuk"
  : assetsPath;
const staticScripts = express.static(scriptsPath);

const stylesPath = shouldLoadFromNodeModules
  ? path.join(rootDir, "./styles")
  : assetsPath;
const staticStyles = express.static(stylesPath);

export const middleware: Middlewares = [
  { handler: generateNonce },
  { handler: generateCSP },
  { handler: staticAssets, route: "/assets" },
  { handler: staticScripts, route: "/assets/scripts" },
  { handler: staticStyles, route: "/assets/styles" },
];

export const unitTestMiddleware: Middlewares = [
  { handler: staticAssets, route: "/assets" },
  { handler: staticScripts, route: "/assets/scripts" },
  { handler: staticStyles, route: "/assets/styles" },
];

export const handleErrors: ErrorRequestHandler = (error, _, response, __) => {
  logger.error("Express app error", { error });
  response.status(500);
  response.render("error.njk");
};
