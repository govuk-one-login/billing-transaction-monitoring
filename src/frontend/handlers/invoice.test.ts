import supertest from "supertest";
import { ConfigElements } from "../../shared/constants";
import { fetchS3, getConfig } from "../../shared/utils";
import { app } from "../app";
import { initApp } from "../init-app";
import { statusLabels } from "../utils";
import { unitTestMiddleware } from "../middleware";

jest.mock("../../shared/utils");
const mockedFetchS3 = fetchS3 as jest.Mock;
const mockedGetConfig = getConfig as jest.Mock;

describe("invoice handler", () => {
  let givenContractsConfig: any;
  let givenServicesConfig: any;
  let givenExtractResults;

  beforeEach(() => {
    initApp(app, unitTestMiddleware);

    process.env = {
      STORAGE_BUCKET: "given storage bucket",
    };

    givenContractsConfig = [
      { id: "1", name: "C01234", vendor_id: "vendor_testvendor1" },
    ];
    givenServicesConfig = [
      {
        vendor_name: "Vendor One",
        vendor_id: "vendor_testvendor1",
        service_name: "Passport check",
        service_regex: "Passport.*check",
        event_name: "VENDOR_1_EVENT_1",
        contract_id: "1",
      },
      {
        vendor_name: "Vendor One",
        vendor_id: "vendor_testvendor1",
        service_name: "ABC Charge",
        service_regex: "ABC.*Charge",
        event_name: "VENDOR_1_EVENT_2",
        contract_id: "1",
      },
    ];

    mockedGetConfig.mockImplementation((fileName) =>
      fileName === ConfigElements.services
        ? givenServicesConfig
        : givenContractsConfig
    );
  });

  test("Page displays Invoice above threshold banner text and three line items in the Reconciliation Table for Mar 2023", async () => {
    // The single line item that's over threshold should cause the error banner
    // even though the other line items are within or below threshold.
    givenExtractResults = `{"vendor_id":"vendor_testvendor1","vendor_name":"Vendor One","service_name":"Passport check","month":"03","year":"2023","contract_id":"1","contract_name":"C01234","billing_price_formatted":"293.19","transaction_price_formatted":"266.82","price_difference":"26.37","billing_quantity":"73298", "transaction_quantity":"66706", "quantity_difference":"6,592.00","billing_amount_with_tax":"351.83","price_difference_percentage":"9.8814"}
    {"vendor_id":"vendor_testvendor1","vendor_name":"Vendor One","service_name":"Passport check","month":"03","year":"2023","contract_id":"1","contract_name":"C01234","billing_price_formatted":"3,850.16","transaction_price_formatted":"3,854.92","price_difference":"-4.76","billing_quantity":"11324", "transaction_quantity":"11338", "quantity_difference":"-14","billing_amount_with_tax":"4,620.19","price_difference_percentage":"-1.1235"}
    {"vendor_id":"vendor_testvendor1","vendor_name":"Vendor One","service_name":"Passport check","month":"03","year":"2023","contract_id":"1","contract_name":"C01234","billing_price_formatted":"100","transaction_price_formatted":"100","price_difference":"0.00","billing_quantity":"100", "transaction_quantity":"100", "quantity_difference":"100.00","billing_amount_with_tax":"110","price_difference_percentage":"0"}`;
    mockedFetchS3.mockResolvedValue(givenExtractResults);

    const request = supertest(app);
    const response = await request.get("/contracts/1/invoices/2023-03");
    expect(response.status).toBe(200);
    expect(response.text).toContain("Vendor One Mar 2023 Invoice"); // page heading
    expect(response.text).toContain("Invoice above threshold"); // banner
    expect(response.text).toContain("9.8814"); // percentage discrepancy in table
    expect(response.text).toContain(
      statusLabels.STATUS_LABEL_ABOVE_THRESHOLD.message
    ); // status in table for Above Threshold
    expect(response.text).toContain("-1.1235"); // percentage discrepancy in table
    expect(response.text).toContain(
      statusLabels.STATUS_LABEL_BELOW_THRESHOLD.message
    ); // status in table for Below Threshold
    expect(response.text).toContain("0"); // percentage discrepancy in table
    expect(response.text).toContain(
      statusLabels.STATUS_LABEL_WITHIN_THRESHOLD.message
    ); // status in table for Within Threshold
    expect(response.text).toMatchSnapshot();
  });
});
