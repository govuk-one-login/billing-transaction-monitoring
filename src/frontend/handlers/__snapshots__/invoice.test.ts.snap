// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invoice handler Page displays Invoice above threshold banner text and three line items in the Reconciliation Table for Mar 2023 1`] = `
"<!DOCTYPE html>
<html lang="en" class="govuk-template ">
  <head>
    <meta charset="utf-8">
    <title>
  
  Vendor One Mar 2023 Invoice -
  Billings and Reconciliation
</title>
    <meta name="viewport" content="width=device-width, initial-scale=1, viewport-fit=cover">
    <meta name="theme-color" content="#0b0c0c"> 
    
    <meta http-equiv="X-UA-Compatible" content="IE=edge">

    
      <link rel="shortcut icon" sizes="16x16 32x32 48x48" href="/assets/images/favicon.ico" type="image/x-icon">
      <link rel="mask-icon" href="/assets/images/govuk-mask-icon.svg" color="#0b0c0c"> 
      <link rel="apple-touch-icon" sizes="180x180" href="/assets/images/govuk-apple-touch-icon-180x180.png">
      <link rel="apple-touch-icon" sizes="167x167" href="/assets/images/govuk-apple-touch-icon-167x167.png">
      <link rel="apple-touch-icon" sizes="152x152" href="/assets/images/govuk-apple-touch-icon-152x152.png">
      <link rel="apple-touch-icon" href="/assets/images/govuk-apple-touch-icon.png">
    

    
  <link href="/assets/styles/all.css" rel="stylesheet">


    
    
  </head>
  <body class="govuk-template__body ">
    <script>document.body.className = ((document.body.className) ? document.body.className + ' js-enabled' : 'js-enabled');</script>
    

    
      <a href="#main-content" class="govuk-skip-link" data-module="govuk-skip-link">Skip to main content</a>

    

    
      

<header class="govuk-header " role="banner" data-module="govuk-header">
  <div class="govuk-header__container govuk-width-container">
    <div class="govuk-header__logo">
      <a href="/" class="govuk-header__link govuk-header__link--homepage">
        <span class="govuk-header__logotype">
          <!--[if gt IE 8]><!-->
          <svg
            aria-hidden="true"
            focusable="false"
            class="govuk-header__logotype-crown"
            xmlns="http://www.w3.org/2000/svg"
            viewBox="0 0 132 97"
            height="30"
            width="36"
          >
            <path
              fill="currentColor" fill-rule="evenodd"
              d="M25 30.2c3.5 1.5 7.7-.2 9.1-3.7 1.5-3.6-.2-7.8-3.9-9.2-3.6-1.4-7.6.3-9.1 3.9-1.4 3.5.3 7.5 3.9 9zM9 39.5c3.6 1.5 7.8-.2 9.2-3.7 1.5-3.6-.2-7.8-3.9-9.1-3.6-1.5-7.6.2-9.1 3.8-1.4 3.5.3 7.5 3.8 9zM4.4 57.2c3.5 1.5 7.7-.2 9.1-3.8 1.5-3.6-.2-7.7-3.9-9.1-3.5-1.5-7.6.3-9.1 3.8-1.4 3.5.3 7.6 3.9 9.1zm38.3-21.4c3.5 1.5 7.7-.2 9.1-3.8 1.5-3.6-.2-7.7-3.9-9.1-3.6-1.5-7.6.3-9.1 3.8-1.3 3.6.4 7.7 3.9 9.1zm64.4-5.6c-3.6 1.5-7.8-.2-9.1-3.7-1.5-3.6.2-7.8 3.8-9.2 3.6-1.4 7.7.3 9.2 3.9 1.3 3.5-.4 7.5-3.9 9zm15.9 9.3c-3.6 1.5-7.7-.2-9.1-3.7-1.5-3.6.2-7.8 3.7-9.1 3.6-1.5 7.7.2 9.2 3.8 1.5 3.5-.3 7.5-3.8 9zm4.7 17.7c-3.6 1.5-7.8-.2-9.2-3.8-1.5-3.6.2-7.7 3.9-9.1 3.6-1.5 7.7.3 9.2 3.8 1.3 3.5-.4 7.6-3.9 9.1zM89.3 35.8c-3.6 1.5-7.8-.2-9.2-3.8-1.4-3.6.2-7.7 3.9-9.1 3.6-1.5 7.7.3 9.2 3.8 1.4 3.6-.3 7.7-3.9 9.1zM69.7 17.7l8.9 4.7V9.3l-8.9 2.8c-.2-.3-.5-.6-.9-.9L72.4 0H59.6l3.5 11.2c-.3.3-.6.5-.9.9l-8.8-2.8v13.1l8.8-4.7c.3.3.6.7.9.9l-5 15.4v.1c-.2.8-.4 1.6-.4 2.4 0 4.1 3.1 7.5 7 8.1h.2c.3 0 .7.1 1 .1.4 0 .7 0 1-.1h.2c4-.6 7.1-4.1 7.1-8.1 0-.8-.1-1.7-.4-2.4V34l-5.1-15.4c.4-.2.7-.6 1-.9zM66 92.8c16.9 0 32.8 1.1 47.1 3.2 4-16.9 8.9-26.7 14-33.5l-9.6-3.4c1 4.9 1.1 7.2 0 10.2-1.5-1.4-3-4.3-4.2-8.7L108.6 76c2.8-2 5-3.2 7.5-3.3-4.4 9.4-10 11.9-13.6 11.2-4.3-.8-6.3-4.6-5.6-7.9 1-4.7 5.7-5.9 8-.5 4.3-8.7-3-11.4-7.6-8.8 7.1-7.2 7.9-13.5 2.1-21.1-8 6.1-8.1 12.3-4.5 20.8-4.7-5.4-12.1-2.5-9.5 6.2 3.4-5.2 7.9-2 7.2 3.1-.6 4.3-6.4 7.8-13.5 7.2-10.3-.9-10.9-8-11.2-13.8 2.5-.5 7.1 1.8 11 7.3L80.2 60c-4.1 4.4-8 5.3-12.3 5.4 1.4-4.4 8-11.6 8-11.6H55.5s6.4 7.2 7.9 11.6c-4.2-.1-8-1-12.3-5.4l1.4 16.4c3.9-5.5 8.5-7.7 10.9-7.3-.3 5.8-.9 12.8-11.1 13.8-7.2.6-12.9-2.9-13.5-7.2-.7-5 3.8-8.3 7.1-3.1 2.7-8.7-4.6-11.6-9.4-6.2 3.7-8.5 3.6-14.7-4.6-20.8-5.8 7.6-5 13.9 2.2 21.1-4.7-2.6-11.9.1-7.7 8.8 2.3-5.5 7.1-4.2 8.1.5.7 3.3-1.3 7.1-5.7 7.9-3.5.7-9-1.8-13.5-11.2 2.5.1 4.7 1.3 7.5 3.3l-4.7-15.4c-1.2 4.4-2.7 7.2-4.3 8.7-1.1-3-.9-5.3 0-10.2l-9.5 3.4c5 6.9 9.9 16.7 14 33.5 14.8-2.1 30.8-3.2 47.7-3.2z"
            ></path>
          </svg>
          <!--<![endif]-->
          <!--[if IE 8]>
          <img src="/assets/images/govuk-logotype-crown.png" class="govuk-header__logotype-crown-fallback-image" width="36" height="32" alt="">
          <![endif]-->
          <span class="govuk-header__logotype-text">
            GOV.UK
          </span>
        </span>
        
      </a>
    </div>
    
  </div>
</header>

    

    
      <div class="govuk-width-container ">
        
  <div class="govuk-breadcrumbs">
  <ol class="govuk-breadcrumbs__list">
  
    
    <li class="govuk-breadcrumbs__list-item">
      <a class="govuk-breadcrumbs__link" href="/">Billings and reconciliation</a>
    </li>
    
  
    
    <li class="govuk-breadcrumbs__list-item">
      <a class="govuk-breadcrumbs__link" href="/contracts">Contracts</a>
    </li>
    
  
    
    <li class="govuk-breadcrumbs__list-item">
      <a class="govuk-breadcrumbs__link" href="/contracts/1/invoices">C01234 - Vendor One</a>
    </li>
    
  
  </ol>
</div>


        <main class="govuk-main-wrapper " id="main-content" role="main">
          
  <h1 class="govuk-heading-xl">Vendor One Mar 2023 Invoice</h1>
  
<div class="govuk-panel govuk-panel--confirmation error">
  <h1 class="govuk-panel__title">
    Invoice above threshold
  </h1>
  
  <div class="govuk-panel__body">
    Reference number
    <br>
    <strong>C01234</strong>
  </div>
  
</div>

  <div class="govuk-inset-text">
    


<div class="govuk-accordion" data-module="govuk-accordion" id="accordion-default">
  
    
      <div class="govuk-accordion__section">
        <div class="govuk-accordion__section-header">
          <h2 class="govuk-accordion__section-heading">
            <span class="govuk-accordion__section-button" id="accordion-default-heading-1">
              Status guidance
            </span>
          </h2>
          
        </div>
        <div id="accordion-default-content-1" class="govuk-accordion__section-content" aria-labelledby="accordion-default-heading-1">
          
            
  <p class='govuk-body'>Common banner states:</p>
  <ul class='govuk-list govuk-list--bullet'>
    <li>
      Invoice within threshold (green) - the discrepancy between measured event data and the invoice is within the threshold tolerance of -1% and 1%. This invoice can be paid.
    </li>
    <li>
      Invoice above threshold (red) - the discrepancy between measured event data and the invoice is above the threshold of 1%. The vendor may have overcharged within the invoice. Contact the Data &amp; Analytics team to confirm measured data and the vendor to confirm the invoice.
    </li>
    <li>
      Invoice below threshold (blue) - the discrepancy between measured event data and the invoice is below the threshold of -1%. The vendor may have undercharged within the invoice. Contact the Data &amp; Analytics team to confirm measured data and the vendor to confirm the invoice.
    </li>
  </ul>
  <p class='govuk-body'>Other banner states (grey):</p>
  <ul class='govuk-list govuk-list--bullet'>
    <li>Invoice and events missing - There is no invoices or event data for this month.</li>
    <li>
      Invoice missing - An event has been measured for this month, but there has been no corresponding invoice received.
    </li>
    <li>
      Events missing - An invoice has been received for this month that includes at least one service that there has been no measured data.
    </li>
    <li>
      Unable to find rate - An error has occurred. An invoice has been received that includes at least one service that does not have a corresponding rate per transaction configured. Contact technical support to resolve this error.
    </li>
    <li>
      Invoice has unexpected charge - Event processing concludes there should be no charge for a particular service for a month (i.e. if all events had a charge Â£0.00), but the invoice charge is non-zero. In this case, event processing cannot compute a percentage discrepancy, as it would be infinite.
    </li>
  </ul>

          
        </div>
      </div>
    
  
</div>

  </div>
  <table class="govuk-table">
    <caption class="govuk-table__caption govuk-table__caption--m">Reconciliation</caption>
    <thead class="govuk-table__head">
      <tr class="govuk-table__row">
        <th class="govuk-table__header" scope="col">Line Item</th>
        <th class="govuk-table__header" scope="col">Quantity Discrepancy</th>
        <th class="govuk-table__header" scope="col">Price Discrepancy</th>
        <th class="govuk-table__header" scope="col">Percentage Discrepancy</th>
        <th class="govuk-table__header" scope="col">Status</th>
      </tr>
    </thead>
    
      <tbody class="govuk-table__body">
        <tr class="govuk-table__row">
          <td class="govuk-table__cell">Passport check</td>
          <td class="govuk-table__cell">6,592.00</td>
          <td class="govuk-table__cell">26.37</td>
          <td class="govuk-table__cell">9.8814%</td>
          <td class="govuk-table__cell">
            <strong class="govuk-tag govuk-tag--red">
  Above Threshold
</strong>

          </td>
        </tr>
      </tbody>
    
      <tbody class="govuk-table__body">
        <tr class="govuk-table__row">
          <td class="govuk-table__cell">Passport check</td>
          <td class="govuk-table__cell">-14</td>
          <td class="govuk-table__cell">-4.76</td>
          <td class="govuk-table__cell">-1.1235%</td>
          <td class="govuk-table__cell">
            <strong class="govuk-tag govuk-tag--blue">
  Below Threshold
</strong>

          </td>
        </tr>
      </tbody>
    
      <tbody class="govuk-table__body">
        <tr class="govuk-table__row">
          <td class="govuk-table__cell">Passport check</td>
          <td class="govuk-table__cell">100.00</td>
          <td class="govuk-table__cell">0.00</td>
          <td class="govuk-table__cell">0%</td>
          <td class="govuk-table__cell">
            <strong class="govuk-tag govuk-tag--green">
  Within Threshold
</strong>

          </td>
        </tr>
      </tbody>
    
  </table>
  <table class="govuk-table">
    <caption class="govuk-table__caption govuk-table__caption--m">Quantity (events)</caption>
    <thead class="govuk-table__head">
      <tr class="govuk-table__row">
        <th class="govuk-table__header" scope="col">Line Item</th>
        <th class="govuk-table__header" scope="col">Invoiced Quantity</th>
        <th class="govuk-table__header" scope="col">Measured Quantity</th>
        <th class="govuk-table__header" scope="col">Quantity Discrepancy</th>
      </tr>
    </thead>
    
      <tbody class="govuk-table__body">
        <tr class="govuk-table__row">
          <td class="govuk-table__cell">Passport check</td>
          <td class="govuk-table__cell">73298</td>
          <td class="govuk-table__cell">66706</td>
          <td class="govuk-table__cell">6,592.00</td>
        </tr>
      </tbody>
    
      <tbody class="govuk-table__body">
        <tr class="govuk-table__row">
          <td class="govuk-table__cell">Passport check</td>
          <td class="govuk-table__cell">11324</td>
          <td class="govuk-table__cell">11338</td>
          <td class="govuk-table__cell">-14</td>
        </tr>
      </tbody>
    
      <tbody class="govuk-table__body">
        <tr class="govuk-table__row">
          <td class="govuk-table__cell">Passport check</td>
          <td class="govuk-table__cell">100</td>
          <td class="govuk-table__cell">100</td>
          <td class="govuk-table__cell">100.00</td>
        </tr>
      </tbody>
    
  </table>
  <table class="govuk-table">
    <caption class="govuk-table__caption govuk-table__caption--m">Price</caption>
    <thead class="govuk-table__head">
      <tr class="govuk-table__row">
        <th class="govuk-table__header" scope="col">Line Item</th>
        <th class="govuk-table__header" scope="col">Invoiced Price</th>
        <th class="govuk-table__header" scope="col">Measured Price</th>
        <th class="govuk-table__header" scope="col">Price Discrepancy</th>
      </tr>
    </thead>
    
      <tbody class="govuk-table__body">
        <tr class="govuk-table__row">
          <td class="govuk-table__cell">Passport check</td>
          <td class="govuk-table__cell">293.19</td>
          <td class="govuk-table__cell">266.82</td>
          <td class="govuk-table__cell">26.37</td>
        </tr>
      </tbody>
    
      <tbody class="govuk-table__body">
        <tr class="govuk-table__row">
          <td class="govuk-table__cell">Passport check</td>
          <td class="govuk-table__cell">3,850.16</td>
          <td class="govuk-table__cell">3,854.92</td>
          <td class="govuk-table__cell">-4.76</td>
        </tr>
      </tbody>
    
      <tbody class="govuk-table__body">
        <tr class="govuk-table__row">
          <td class="govuk-table__cell">Passport check</td>
          <td class="govuk-table__cell">100</td>
          <td class="govuk-table__cell">100</td>
          <td class="govuk-table__cell">0.00</td>
        </tr>
      </tbody>
    
  </table>
  <table class="govuk-table">
    <caption class="govuk-table__caption govuk-table__caption--m">Measured (events)</caption>
    <thead class="govuk-table__head">
      <tr class="govuk-table__row">
        <th class="govuk-table__header" scope="col">Line Item</th>
        <th class="govuk-table__header" scope="col">Quantity</th>
      </tr>
    </thead>
    
      <tbody class="govuk-table__body">
        <tr class="govuk-table__row">
          <td class="govuk-table__cell">Passport check</td>
          <td class="govuk-table__cell">66706</td>
        </tr>
      </tbody>
    
      <tbody class="govuk-table__body">
        <tr class="govuk-table__row">
          <td class="govuk-table__cell">Passport check</td>
          <td class="govuk-table__cell">11338</td>
        </tr>
      </tbody>
    
      <tbody class="govuk-table__body">
        <tr class="govuk-table__row">
          <td class="govuk-table__cell">Passport check</td>
          <td class="govuk-table__cell">100</td>
        </tr>
      </tbody>
    
  </table>

        </main>
      </div>
    

    
      <footer class="govuk-footer " role="contentinfo">
  <div class="govuk-width-container ">
    
    <div class="govuk-footer__meta">
      <div class="govuk-footer__meta-item govuk-footer__meta-item--grow">
        
        <svg
          aria-hidden="true"
          focusable="false"
          class="govuk-footer__licence-logo"
          xmlns="http://www.w3.org/2000/svg"
          viewBox="0 0 483.2 195.7"
          height="17"
          width="41"
        >
          <path
            fill="currentColor"
            d="M421.5 142.8V.1l-50.7 32.3v161.1h112.4v-50.7zm-122.3-9.6A47.12 47.12 0 0 1 221 97.8c0-26 21.1-47.1 47.1-47.1 16.7 0 31.4 8.7 39.7 21.8l42.7-27.2A97.63 97.63 0 0 0 268.1 0c-36.5 0-68.3 20.1-85.1 49.7A98 98 0 0 0 97.8 0C43.9 0 0 43.9 0 97.8s43.9 97.8 97.8 97.8c36.5 0 68.3-20.1 85.1-49.7a97.76 97.76 0 0 0 149.6 25.4l19.4 22.2h3v-87.8h-80l24.3 27.5zM97.8 145c-26 0-47.1-21.1-47.1-47.1s21.1-47.1 47.1-47.1 47.2 21 47.2 47S123.8 145 97.8 145"
          />
        </svg>
        <span class="govuk-footer__licence-description">
          
            All content is available under the
            <a
              class="govuk-footer__link"
              href="https://www.nationalarchives.gov.uk/doc/open-government-licence/version/3/"
              rel="license"
            >Open Government Licence v3.0</a>, except where otherwise stated
          
        </span>
      </div>
      <div class="govuk-footer__meta-item">
        <a
          class="govuk-footer__link govuk-footer__copyright-logo"
          href="https://www.nationalarchives.gov.uk/information-management/re-using-public-sector-information/uk-government-licensing-framework/crown-copyright/"
        >Â© Crown copyright</a>
      </div>
    </div>
  </div>
</footer>

    

    
  <script src="assets/scripts/all.js"></script>
  <script>window.GOVUKFrontend.initAll();</script>
  
  

  </body>
</html>
"
`;
